# Vaccine Efficacy

Source : https://sachin-date.medium.com/how-to-estimate-vaccine-efficacy-using-a-logistic-regression-model-121f9ca5a9d8

$Vaccine Efficacy (VE)  = 1 - Incident Rate Ratio (IRR)$  
$IRR = \frac{p_{vacinated}} {p_{placebo}}$

In [2]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [3]:
# df = pd.read_csv('https://gist.githubusercontent.com/sachinsdate/2a6dc052b3c416e24a4d04c468358c44/raw/376939c4a1e8205f6bda03d0415b9758986ff693/vaccine_trial_simulation_study.csv')
df = pd.read_csv('vaccine_trial_simulation_study.csv')

In [4]:
df.head()

INTERVAL_BETWEEN_DOSES  VACCINATED  INFECTED
0                       0           1         1
1                       0           1         1
2                       0           1         1
3                       0           1         1
4                       0           1         1

## Analysis
1 Standard inference with statistical test  
2 Standard inference with Statsmodel  
3 Bootstraping inference  

### Basic Summary

In [1]:
pd.crosstab(index=df['VACCINATED'], columns=df['INFECTED'])

NameError: name 'pd' is not defined

In [ ]:
pd.crosstab(index=df['VACCINATED'], columns=df['INFECTED'], margins=True)

In [11]:
pd.crosstab(index=df['VACCINATED'], columns=df['INFECTED'], margins=True, normalize=True)

INFECTED           0         1     All
VACCINATED                            
0           0.460250  0.046250  0.5065
1           0.490917  0.002583  0.4935
All         0.951167  0.048833  1.0000

### 1) Test if the `INFECTED` independence of `VACCINATED`  

#### 1.1) $\chi^2$ test of independence, library `Scipy`  
reference : https://www.jmp.com/en_us/statistics-knowledge-portal/chi-square-test.html  
reference : https://stats.stackexchange.com/questions/110718/chi-squared-test-with-scipy-whats-the-difference-between-chi2-contingency-and  

In [22]:
from scipy.stats import chi2_contingency

In [23]:
contingency_tbl = pd.crosstab(index=df['VACCINATED'], columns=df['INFECTED'])

In [26]:
chi, p, ddof, expected = chi2_contingency(contingency_tbl)
expected

array([[5781.191,  296.809],
       [5632.809,  289.191]])

In [27]:
print(p)

1.1586205594709897e-105


With p = 1.15 e-105, chance of repeate testing and find the same result : dependence between 2 categorical variables

#### 1.2) $\chi^2$ test of independence, library `statsmodels.`
reference : https://www.statsmodels.org/stable/contingency_tables.html

In [16]:
contingency_tbl = pd.crosstab(index=df['VACCINATED'], columns=df['INFECTED'])
# Convert numpy table to Statsmodel contingency table
cont_tbl_sm = sm.stats.Table(contingency_tbl)

In [19]:
# original contingency table
cont_tbl_sm.table_orig

INFECTED       0    1
VACCINATED           
0           5523  555
1           5891   31

In [28]:
# Expected contingency table if factor are independence
cont_tbl_sm.fittedvalues

INFECTED           0        1
VACCINATED                   
0           5781.191  296.809
1           5632.809  289.191

In [37]:
test_chi_indep = cont_tbl_sm.test_nominal_association()
test_chi_indep.pvalue

0.0

### 2) Test if proportion of `INFECTED` differ between `VACCINATED` group

#### 2.1) Sampling distribution test with $Z-test$ , library `scipy`

reference : https://online.stat.psu.edu/stat415/lesson/9/9.4

reference : https://www.statsmodels.org/stable/generated/statsmodels.stats.proportion.test_proportions_2indep.html#statsmodels.stats.proportion.test_proportions_2indep

`Assumption CTL` difference of sampling proportion is *Normal*
Hypothesis Test  
$H_0$ : $P_{vac} = P_{not vac.}$, $P_{vac} - P_{not vac.} = 0$  
$H_A$ : $P_{vac} \neq P_{not vac.}$

In [57]:
df_vac = df[df['VACCINATED'] == 1]
df_not_vac = df[df['VACCINATED'] != 1]
n_vac = df_vac.shape[0]
n_not_vac = df_not_vac.shape[0]
n_combine = df.shape[0]
p_vac = df_vac['INFECTED'].sum()/n_vac
p_not_vac = df_not_vac['INFECTED'].sum()/n_not_vac
p_combine = df['INFECTED'].sum()/n_combine
se_combine = np.sqrt((p_combine*(1-p_combine))*(1/n_not_vac + 1/n_vac))

In [58]:
print(f'{p_vac}, {n_vac}'); print(f'{p_not_vac}, {n_not_vac}'); print(f'{p_combine}, {n_combine}')

0.005234718000675448, 5922
0.09131293188548864, 6078
0.04883333333333333, 12000


In [65]:
z_test_score = ((p_not_vac - p_vac) - (0))/se_combine

In [66]:
import scipy as sp

print(z_test_score)
p_value = sp.stats.norm.sf(z_test_score)
print(p_value)

21.874124374627076
2.291116351867227e-106


#### 2.2) Sampling distribution test with $Z-test$ , library `statsmodels`
reference : https://sonalake.com/latest/hypothesis-testing-of-proportion-based-samples/

In [67]:
from statsmodels.stats.proportion import proportions_ztest
import numpy as np

vac_infected, vac_size = (df_vac['INFECTED'].sum(), df_vac.shape[0])
not_vac_infected, not_vac_infected_size = (df_not_vac['INFECTED'].sum(), df_not_vac.shape[0])

# check our sample against Ho for Ha != Ho
total_infected = np.array([not_vac_infected, vac_infected])
total_size = np.array([not_vac_infected_size, vac_size])

# note, no need for a Ho value here - it's derived from the other parameters
stat, p_value = proportions_ztest(count=total_infected, nobs=total_size,  alternative='two-sided')

# report
print('z_stat: %0.3f, p_value: %0.3f' % (stat, p_value))

z_stat: 21.874, p_value: 0.000


#### 2.3) Bootstraping / Monte Carlo test  
reference : https://stats.stackexchange.com/questions/395120/proportion-test-z-test-vs-bootstrap-simulation-different-results  

Null pypothesis  
$H_0 : p_{not_vac} = p_{vac}$ or $p_{not_vac} - p_{vac} = 0$

In [11]:
pd.crosstab(index=df['VACCINATED'], columns=df['INFECTED'], margins=True, normalize='index')

INFECTED           0         1
VACCINATED                    
0           0.908687  0.091313
1           0.994765  0.005235
All         0.951167  0.048833

Hypothesis test with `Mote Carlo` method, assume Binomial Distribution

In [34]:
from numpy.random import binomial

p_0 = df['INFECTED'].mean()
p_a = df[df['VACCINATED'] == 1]['INFECTED'].mean()
n_size = df.shape[0]

print(f'p-value : {sum((np.random.binomial(n_size, p_0, 10000)/n_size) < p_a)/10000}')

p-value : 0.0


Confidence interval of difference proportion with `Bootstrap`

In [36]:
def diff_prop(df):
    """calculate difference proportion infected"""
    vac = df[df['VACCINATED'] == 1]
    not_vac = df[df['VACCINATED'] != 1]
    n_vac_infected = vac['INFECTED'].sum()
    n_not_vac_infected = not_vac['INFECTED'].sum()

    return (n_not_vac_infected/not_vac.shape[0]) - (n_vac_infected/vac.shape[0])

In [37]:
bs_replicate = np.empty(10000)
for i in range(10000):
    bs_df = df.sample(n=df.shape[0], replace=True)
    bs_replicate[i] = diff_prop(bs_df)

In [38]:
# conf interval 5% of difference of proportion
np.percentile(bs_replicate, [5, 95])

array([0.07990947, 0.09236965])

## Logistic Regression for Vaccine Efficacy
library `statsmodels`

Reference : https://sachin-date.medium.com/how-to-estimate-vaccine-efficacy-using-a-logistic-regression-model-121f9ca5a9d8

In [4]:
import statsmodels.formula.api as smf

In [7]:
logit_model = smf.logit('INFECTED ~ INTERVAL_BETWEEN_DOSES + VACCINATED', data = df)
logit_result = logit_model.fit()
logit_result.summary()

Optimization terminated successfully.
         Current function value: 0.170907
         Iterations 9


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               INFECTED   No. Observations:                12000
Model:                          Logit   Df Residuals:                    11997
Method:                           MLE   Df Model:                            2
Date:                Sat, 15 May 2021   Pseudo R-squ.:                  0.1238
Time:                        09:26:02   Log-Likelihood:                -2050.9
converged:                       True   LL-Null:                       -2340.8
Covariance Type:            nonrobust   LLR p-value:                1.254e-126
==========================================================================================
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -2.3154      0.063    -36.816      0.000      -2.439      -2.192
INTERVAL_BETWEEN_DOSES     0.0347      0.086      0.401      0.688      -0.135       0.204
VACCINATED                -2.9491      0.186    -15.898      0.000      -3.313      -2.585
==========================================================================================
"""

### Key Points
- `Log-Likelihood Ratio` is 1.25e-126 (which is essentially zero) indicating that the model does indeed fit better than a Null model  
- `psuedo R-squared` is only 12.38% indicating a poor fit. We might want to experiment with a Poisson, a Generalized Poisson or a Negative Binomial  
- `p-value` of the coefficient of the `INTERVAL_BETWEEN_DOSES` variable is 0.688, which means that we cannot say at even a 40% confidence level that the coefficient of `INTERVAL_BETWEEN_DOSES` is really any different from 0.

`Question` : the coefficent if split between `Vaccinated` and `Non-Vaccinated` data group will be the same?

In [8]:
logit_vac = smf.logit('INFECTED ~ INTERVAL_BETWEEN_DOSES', data = df[df['VACCINATED'] == 1])
logit_vac_result = logit_vac.fit()
logit_non_vac = smf.logit('INFECTED ~ INTERVAL_BETWEEN_DOSES', data = df[df['VACCINATED'] == 0])
logit_non_vac_result = logit_non_vac.fit()

Optimization terminated successfully.
         Current function value: 0.031917
         Iterations 10
Optimization terminated successfully.
         Current function value: 0.305456
         Iterations 6


In [9]:
logit_vac_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               INFECTED   No. Observations:                 5922
Model:                          Logit   Df Residuals:                     5920
Method:                           MLE   Df Model:                            1
Date:                Sat, 15 May 2021   Pseudo R-squ.:                 0.02442
Time:                        09:28:29   Log-Likelihood:                -189.01
converged:                       True   LL-Null:                       -193.74
Covariance Type:            nonrobust   LLR p-value:                  0.002098
==========================================================================================
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -4.8203      0.205    -23.520      0.000      -5.222      -4.419
INTERVAL_BETWEEN_DOSES    -1.2114      0.430     -2.815      0.005      -2.055      -0.368
==========================================================================================
"""

In [10]:
logit_non_vac_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               INFECTED   No. Observations:                 6078
Model:                          Logit   Df Residuals:                     6076
Method:                           MLE   Df Model:                            1
Date:                Sat, 15 May 2021   Pseudo R-squ.:               0.0003578
Time:                        09:28:55   Log-Likelihood:                -1856.6
converged:                       True   LL-Null:                       -1857.2
Covariance Type:            nonrobust   LLR p-value:                    0.2490
==========================================================================================
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -2.3508      0.065    -36.289      0.000      -2.478      -2.224
INTERVAL_BETWEEN_DOSES     0.1028      0.089      1.152      0.249      -0.072       0.278
==========================================================================================
"""